### PR TITLE
DDF-4202 AbstractCswSource to transform Query Filter Properties based on typeName

### PR DESCRIPTION
**Pull request by lamhuy**
#### What does this PR do?
AbstractCswSource to transform Query Filter Properties based on typeName
CSW Source need to be able to transform from internal query representation to another given query representation during federation. 
As part of a source configuration, typeName is provided. This should be used to look up the appropriate QueryFilterTransformer.

#### Who is reviewing it? 
@codice/ogc 
@bdeining 
@millerw8 
@mojogitoverhere 
@AzGoalie 
@troymohl 
@jlcsmith 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@millerw8
@jlcsmith 

#### How should this be tested?
Ensure build and CI run successfully
Upload an image into DDF.
Create a CSW Specification federated source. Leave all default configuration
From UI, perform a query against this federated source and verify the result

#### Any background context you want to provide?
This is an architectural enabler for 
https://codice.atlassian.net/browse/DDF-4204
https://codice.atlassian.net/browse/DDF-4207

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-4202

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/spatial/csw/spatial-csw-transformer/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -150,8 +150,28 @@
...
+    <reference-list id="queryFilterTransformerList"
+      interface="ddf.catalog.transform.QueryFilterTransformer"
+      member-type="service-reference"
+      availability="optional">
+        <reference-listener bind-method="bind"
+          unbind-method="unbind"
+          ref="cswQueryFilterTransformerProvider"/>
+    </reference-list>
+
+    <service ref="cswQueryFilterTransformerProvider" id="cswQueryFilterTransformerProviderService"
```
Comment (Bdthomson) [inline catalog/spatial/csw/spatial-csw-transformer/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Why does this have a ref and an id?

**Comment by paouelle** [comment]
❓ Should the coverage numbers in the pom file should be updated as new code was added.

File: catalog/spatial/csw/spatial-csw-transformer/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (lamhuy) [reply catalog/spatial/csw/spatial-csw-transformer/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: `ref` is referencing to the bean that providing this service, while `id` is the id of the service, which is kinda optional.  http://aries.apache.org/modules/blueprint.html

Comment (Bdthomson) [reply catalog/spatial/csw/spatial-csw-transformer/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Gotcha, I've never put an id on a service.

**Comment by lamhuy** [comment]
build now

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/QueryFilterTransformerProvider.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * <b> This code is experimental. While this interface is functional and tested, it may change or be
+ * removed in a future version of the library. </b>
+ *
+ * <p>Manages a reference list of {@link QueryFilterTransformer}'s by mapping them to the typenames
+ * they apply to.
+ */
+public interface QueryFilterTransformerProvider {
+
+  Optional<QueryFilterTransformer> getTransformer(QName qName);
```
Comment (clockard) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/QueryFilterTransformerProvider.java]: Could use some Javadoc on api methods

**Comment by lamhuy** [comment]
build now

**Comment by lamhuy** [comment]
build now

**Comment by lamhuy** [comment]
build now

**Comment by alanchampion** [comment]
✔️ Hero successful! 
Successfully was able to create CSW self federated source and ingested and retrieved data using the source.

**Pull request merged**